juju/terraform-provider-juju
# [JUJU-3789] Add TEST_CLOUD environment variable for CI.

## Description

Currently we do not distinguish testing scenarios. With this PR we enable tests checking what type of cloud is provided by the controller. The idea is that we can skip those tests requiring specific infrastructure scenarios. Right now we only distinguish between `lxd` and `microk8s` clouds in the environment variable `TEST_CLOUD`.

Additional changes around tests to request specific environments will be addressed in a separated PR.

## Type of change

Please mark if proceeds.


## Environment

- Juju controller version: 2.9.42
- Terraform version: 1.3.7

## QA steps

Run the 

Run the `TestAcc_ResourceApplication_Basic` with the wrong `TEST_CLOUD` value and say how it skips.

```bash
export CONTROLLER=$(juju whoami | yq .Controller)
export JUJU_CONTROLLER_ADDRESSES="$(juju show-controller | yq '.[$CONTROLLER]'.details.\"api-endpoints\" | tr -d "[]' "|tr -d '"'|tr -d '\n')"
export JUJU_USERNAME="$(cat ~/.local/share/juju/accounts.yaml | yq .controllers.$CONTROLLER.user|tr -d '"')"
export JUJU_PASSWORD="$(cat ~/.local/share/juju/accounts.yaml | yq .controllers.$CONTROLLER.password|tr -d '"')"
export JUJU_CA_CERT="$(juju show-controller $(echo $CONTROLLER|tr -d '"') | yq '.[$CONTROLLER]'.details.\"ca-cert\"|tr -d '"'|sed 's/\\n/\n/g')"

 TF_ACC=1 TEST_CLOUD=microk8s go test ./... -v -run="^TestAcc_ResourceApplication_Basic$" 
...
=== RUN   TestAcc_ResourceApplication_Basic
    resource_application_test.go:16: TestAcc_ResourceApplication_Basic only runs with LXD
--- SKIP: TestAcc_ResourceApplication_Basic (0.00s)
PASS
```

Using the `lxd` value, the test should run.

```sh
TF_ACC=1 TEST_CLOUD=lxd go test ./... -v -run="^TestAcc_ResourceApplication_Basic$" 
...
=== RUN   TestAcc_ResourceApplication_Basic
{"level":"trace","time":"2023-05-25T11:58:05+02:00","message":"call expose application [test-app]"}
--- PASS: TestAcc_ResourceApplication_Basic (20.43s)
PASS
```

# Additional notes
NA
